circuit/environment/src/circuit.rs

```diff
@@ -146,6 +147,13 @@ impl Environment for Circuit {
             // Ensure we are not in witness mode.
             if !in_witness.get() {
                 CIRCUIT.with(|circuit| {
+                    // Ensure we do not surpass maximum allowed number of constraints
+                    MAX_NUM_CONSTRAINTS.with(|max_constraints| {
+                        if circuit.borrow().num_constraints() > max_constraints.get() {
+                            Self::halt("Surpassing maximum allowed number of constraints")
+                        }
+                    });
+
```

Comment:
shouldn't this check happen at the end of this function instead? I may be reading this wrong, but is seems to me that having this at the beginning could allow the following to happen:
- the check for max number of constraints passes
- new constraints are introduced, and they breach the limit
- the program is halted only if `enforce` is called again afterwards

Reply:
ah, I thought `enforce` could potentially introduce more than a single constraint, but since it's not the case, it's nice to avoid an off-by-one